### PR TITLE
Fixing DCF logic bugs

### LINKED ISSUE #84: Dcf calculation fails for single spiral arm
When the trajectory is a single spiral arm, then `DcfData.from_traj_voronoi` detects this as a single non-singleton dimension and carries out a 1D calculation. This fails for a spiral trajectory because it covers a 2D k-space. 

Possible solution: Check if points lie on a straight line, otherwise use voronoi
This will still fail for certain trajectories but this then needs to be handled by the user with a case by case solution.

### PR DESCRIPTION
**Pull request by fzimmermann89**
The fix for #84 and related bugs.
First, I added some currently failing tests showing the issues @ckolbPTB noticed.
Next, The fix.

Closes #84
Supersedes #85

### CODE REVIEW

**Comment by fzimmermann89** [comment]
I also found some (mostly irrelevant) other bugs:
-  The 1D calculation failed for less then 3 unique values, this can be tested for with a 3D cartesian with (2,2,2)
- The voronoi-outlier detection had a bug for less then 10 samples.
and added testcases.

**Comment by fzimmermann89** [comment]
..and, I hope, the fixes.

The broadcast logic is now correct: Instead of checking if in one of kx/ky/kz a dimension is 1, now the logic is:
For each of k2,k1,k0 it is checked if more than one of kx,ky,kz change. 

If yes, we need voronoi. 
Otherwise, we use the 1D calculation for the one that changes or ignore the dimension if neither of kx,ky,kz changes.

File: src/mrpro/data/_DcfData.py [hunk 1]
```diff
@@ -58,10 +58,17 @@ def _dcf_1d(traj: torch.Tensor) -> torch.Tensor:
         # we assign the point at xN the area (xN - xN-1) / 2 + (xN - xN-1) / 2.
 
         kernel = torch.tensor([-1 / 2, 0, 1 / 2], dtype=torch.float32, device=traj.device).reshape(1, 1, 3)
-        central_diff = torch.nn.functional.conv1d(traj_sorted[None, None, :], kernel)[0, 0]
-        first = traj_sorted[1] - traj_sorted[0]
-        last = traj_sorted[-1] - traj_sorted[-2]
-        central_diff = torch.cat((first[None], central_diff, last[None]), -1)
+
+        if (elements := len(traj_sorted)) >= 3:
```
Comment (fzimmermann89) [inline src/mrpro/data/_DcfData.py:62]: This is the common case.

The other two cases should realistically not be hit to often:
Only for k-dimensions with less then 3 unique values.

File: src/mrpro/data/_DcfData.py [hunk 2]
```diff
@@ -136,26 +144,35 @@ def from_traj_voronoi(cls, traj: KTrajectory) -> DcfData:
...
         dcfs = []
-        for k in [traj.kz, traj.ky, traj.kx]:
-            non_singleton_dims = tuple(dim for dim in (-1, -2, -3) if k.shape[dim] != 1)
-            if len(non_singleton_dims) == 1:
-                # Found a direction with two singleton dimensions and one non-singleton dimension
+
+        ks = [traj.kz, traj.ky, traj.kx]
+        spatial_dims = (-3, -2, -1)
+        ks_needing_voronoi = set()
+        for dim in spatial_dims:
```
Comment (fzimmermann89) [inline src/mrpro/data/_DcfData.py:154]: This is the main fix

File: src/mrpro/data/_DcfData.py [hunk 3]
```diff
@@ -136,26 +144,35 @@ def from_traj_voronoi(cls, traj: KTrajectory) -> DcfData:
...
+            non_singleton_ks = [ax for ax in ks if ax.shape[dim] != 1]
+            if len(non_singleton_ks) == 1:
+                # Found a dimension only one non-singleton axes
```
Comment (ckolbPTB) [inline src/mrpro/data/_DcfData.py]: Found a dimension WITH only one non-singleton axes

File: src/mrpro/data/_DcfData.py [hunk 4]
```diff
@@ -136,26 +144,35 @@ def from_traj_voronoi(cls, traj: KTrajectory) -> DcfData:
...
                 # We can handle this as a 1D trajectory
-                dcfs.append(smap(DcfData._dcf_1d, k, non_singleton_dims))
-            elif len(non_singleton_dims) == 0:
-                # Found a direction with only singleton dimensions, i.e. constant dcf
-                dcfs.append(torch.ones_like(k))
-            else:
+                dcfs.append(smap(DcfData._dcf_1d, non_singleton_ks.pop(), (dim,)))
+            elif len(non_singleton_ks) > 0:
                 # A Full Dimension needing voronoi
```
Comment (ckolbPTB) [inline src/mrpro/data/_DcfData.py]: A full dimension needing...

File: src/mrpro/data/_DcfData.py [hunk 5]
```diff
@@ -136,26 +144,35 @@ def from_traj_voronoi(cls, traj: KTrajectory) -> DcfData:
...
-                ks_needing_voronoi.append(k)
+                ks_needing_voronoi |= set(non_singleton_ks)
+            else:
+                # A all 1-dimension, we dont need to do anything
```
Comment (ckolbPTB) [inline src/mrpro/data/_DcfData.py]: All dimensions are singleton, we don't need to do anything

File: src/mrpro/data/_DcfData.py [hunk 6]
```diff
@@ -136,26 +144,35 @@ def from_traj_voronoi(cls, traj: KTrajectory) -> DcfData:
...
-            dcfs.append(smap(DcfData._dcf_2d3d_voronoi, torch.stack(ks_needing_voronoi, -4), 4))
-        # Multiply all dcfs together
-        dcf = reduce(torch.mul, dcfs)
+            dcfs.append(smap(DcfData._dcf_2d3d_voronoi, torch.stack(list(ks_needing_voronoi), -4), 4))
+
+        if dcfs:
+            # Multiply all dcfs together
+            dcf = reduce(torch.mul, dcfs)
+        else:
+            # only singleton spatial dimensions
```
Comment (ckolbPTB) [inline src/mrpro/data/_DcfData.py]: **O**nly

**Review by ckolbPTB — APPROVED** [review]
Just a few typos.

**Pull request merged**